runtime/lua/vim/lsp/folding_range.lua

```diff
@@ -0,0 +1,365 @@
+  vim._with({ win = winid }, function()
+    -- Ensuring folding the smallest ranges first is not explicitly supported,
+    -- but since LSP servers often return the largest ranges first,
+    -- we traverse the ranges in reverse order here.
+    --
+    -- If sorting needs to be implemented,
+    -- the most suitable algorithm here is
+    -- [Timesort](https://wikipedia.org/wiki/Timsort).
+    --
+    -- "Hanc marginis exiguitas non caperet."
+    for i = #ranges, 1, -1 do
+      local range = ranges[i]
+      local start_lnum = range.startLine + 1
+      local end_lnum = range.endLine + 1
+      -- Folding a single line is not supported.
+      if start_lnum ~= end_lnum then
+        vim.cmd(start_lnum .. ',' .. end_lnum .. 'foldclose')
+      end
+    end
+  end)
+end
+
+--- Close all {kind} of folds in the the window with {winid}.
+---
+--- To automatically fold imports when opening a file, you can use an autocmd:
+---
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspNotify', {
+---   callback = function(args)
+---     if args.data.method == 'textDocument/didOpen' then
+---       vim.lsp.folding_range.foldclose('imports', vim.fn.bufwinid(args.buf))
+---     end
+---   end,
+--- })
+--- ```
+---
+---@param kind lsp.FoldingRangeKind "comment" | "imports" | "region"
+---@param winid? integer Defaults to the current window.
+function M.foldclose(kind, winid)
+  vim.validate('kind', kind, 'string')
+  vim.validate('winid', winid, 'number', true)
+
+  winid = winid or api.nvim_get_current_win()
+  local bufnr = api.nvim_win_get_buf(winid)
+  local bufstate = bufstates[bufnr]
+  if not bufstate then
+    return
+  end
+  ---@cast bufstate vim.lsp.folding_range.BufState
+
+  -- Schedule `foldclose()` if the buffer is not up-to-date.
+  if bufstate.version ~= util.buf_versions[bufnr] then
+    for _, client in
+      ipairs(vim.lsp.get_clients({ bufnr = bufnr, method = ms.textDocument_foldingRange }))
+    do
+      local request_id = refresh(bufnr, client)
+      api.nvim_create_autocmd('LspRequest', {
+        buffer = bufnr,
+        callback = function(args)
+          -- Do foldclose() if the `refresh()` request above is completed.
+          if args.data.request_id == request_id and args.data.request.type == 'complete' then
+            local ranges = bufstate.kind_ranges[kind] or {}
+            foldclose(winid, ranges)
+            return true
+          end
+        end,
+      })
+    end
+    return
+  end
+
+  local ranges = bufstate.kind_ranges[kind] or {}
+  foldclose(winid, ranges)
+end
+
+---@return string
+function M._foldtext()
+  local bufnr = api.nvim_get_current_buf()
+  local lnum = vim.v.foldstart
+  local row = lnum - 1
+  local bufstate = bufstates[bufnr]
+  if bufstate and bufstate.row_text[row] then
+    return bufstate.row_text[row]
+  end
+  return vim.fn.getline(lnum)
+end
+
+---@param lnum? integer
+---@return string level
+function M._foldexpr(lnum)
+  vim.validate('lnum', lnum, 'number', true)
```

Comment:
This function is pretty hot and is only called internally by vim, so I think this `vim.validate` can be removed.